walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,73 @@
 
 package com.google.sps;
 
+import java.util.Set;
+import java.util.HashSet;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> getComplement(ArrayList<TimeRange> timesToAvoid, int duration) {
+        // get rid of intervals contained within other intervals
+        Collections.sort(timesToAvoid, TimeRange.ORDER_BY_START);
+        int latestSoFar = -1;
+        ArrayList<TimeRange> filteredTimesToAvoid = new ArrayList<>();
+        for(TimeRange tr: timesToAvoid) {
+            if(tr.end() > latestSoFar) {
+                latestSoFar = tr.end();
+                filteredTimesToAvoid.add(tr);
+            }
+        }
+        // get the times in between
+        Collection<TimeRange> res = new ArrayList<>();
+        int curStart = TimeRange.START_OF_DAY;
+        for(TimeRange tr: filteredTimesToAvoid) {
+            if(tr.start() - curStart >= duration) {
+                res.add(TimeRange.fromStartEnd(curStart, tr.start(), false));
+            }
+            curStart = tr.end();
+        }
+        if(TimeRange.END_OF_DAY+1 - curStart >= duration)
+            res.add(TimeRange.fromStartEnd(curStart, TimeRange.END_OF_DAY+1, false));
+        return res;
+    }
+
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        Set<String> mandatoryAttendees = new HashSet<>();
+        for(String attendee: request.getAttendees()) {
+            mandatoryAttendees.add(attendee);
+        }
+        Set<String> optionalAttendees = new HashSet<>();
+        for(String attendee: request.getOptionalAttendees()) {
+            optionalAttendees.add(attendee);
+        }
+        int duration = (int)request.getDuration();
+        if(mandatoryAttendees.size() == 0) {
+            ArrayList<TimeRange> timesToAvoid = new ArrayList<>();
+            for(Event event: events) {
+                if(!Collections.disjoint(event.getAttendees(), optionalAttendees)) {
+                    timesToAvoid.add(event.getWhen());
+                }
+            }
+            return getComplement(timesToAvoid, duration);
```

Comment:
The if branch ends with a "return". In this way, the else branch is redundant. So the code can be
```java

if(mandatoryAttendees.size() == 0) {
     ArrayList<TimeRange> timesToAvoid = new ArrayList<>();
     for(Event event: events) {
          if(!Collections.disjoint(event.getAttendees(), optionalAttendees)) {
              timesToAvoid.add(event.getWhen());
          }
      }
      return getComplement(timesToAvoid, duration);
}

ArrayList<TimeRange> timesToAvoid = new ArrayList<>();
 ...
```

Reply:
Personally I don't mind this redundancy because I think it makes the code clearer. If there are no mandatory employees, do this. Otherwise do that.

Without the else block, the reader would have to deduce on their own that the two blocks are disjoint, which may not be clear at first glance.